**To connect MySQL to Python, you can use the Python MySQL connector module. Here are the steps to follow:**

1. Install the MySQL connector module in Python by running the following command in your command prompt or terminal:
- pip install mysql-connector-python

2. Import the mysql.connector module in your Python script:

- import mysql.connector

3. Establish a connection to your MySQL database by creating a MySQLConnection object and passing in the required parameters such as the host, user, password, and database name. For example:

    mydb = mysql.connector.connect(
        host="localhost",
        user="yourusername",
        password="yourpassword",
        database="mydatabase"
    )
4. Once the connection is established, you can create a cursor object to execute SQL queries. For example:

    mycursor = mydb.cursor()

5. Execute your SQL queries using the execute() method of the cursor object. For example:
    
    mycursor.execute("SELECT * FROM customers")

6. To retrieve the results of your query, you can use the fetchall() or fetchone() methods of the cursor object. For example:

    myresult = mycursor.fetchall()

    for row in myresult:
        print(row)

**Note: Make sure you have the correct credentials and access privileges to connect to your MySQL database.**

In [ ]:
import mysql.connector 

In [2]:
imdb = mysql.connector.connect(
    host='localhost',
    user='root',
    password='cdacdbda',
    database='imdb'
)

In [3]:
cusor_imdb=imdb.cursor()

In [4]:
cusor_imdb.execute("select * from movie limit 5")

In [5]:
x = cusor_imdb.fetchall()

In [8]:
for _ in x:
    print(_)

('tt0012494', 'Der müde Tod', 2017, datetime.date(2017, 6, 9), 97, 'Germany', '$ 12156', 'German', 'Decla-Bioscop AG')
('tt0038733', 'A Matter of Life and Death', 2017, datetime.date(2017, 12, 8), 104, 'UK', '$ 124241', 'English, French, Russian', 'The Archers')
('tt0060908', 'Le roi de coeur', 2018, datetime.date(2018, 6, 8), 102, 'France, Italy', '$ 18130', 'French, English, German', 'Fildebroc')
('tt0069049', 'The Other Side of the Wind', 2018, datetime.date(2018, 11, 2), 122, 'France, Iran, USA', None, 'English, German', 'Royal Road Entertainment')
('tt0071145', 'Ankur', 2019, datetime.date(2019, 1, 2), 131, 'India', None, 'Hindi', 'Blaze Film Enterprises')


In [22]:
import pandas as pd

In [109]:
def Query(a): 
    '''
    This Function runs the SQL Query and Returns a Markdown
    '''
    cusor_imdb=imdb.cursor()
    cusor_imdb.execute(a)
    x = cusor_imdb.fetchall()
    sql_Data=pd.DataFrame(x)
    sql_Data.columns=cusor_imdb.column_names
    print(sql_Data.to_markdown())

In [112]:
def Query_df(a):
    '''
    This Function runs the SQL Query and Returns a Dataframe
    '''
    cusor_imdb=imdb.cursor()
    cusor_imdb.execute(a)
    x = cusor_imdb.fetchall()
    sql_Data=pd.DataFrame(x)
    sql_Data.columns=cusor_imdb.column_names
    return sql_Data.head()

In [24]:
Query("Select * from movie")

id                       title  year date_published  duration  \
0  tt0012494                Der müde Tod  2017     2017-06-09        97   
1  tt0038733  A Matter of Life and Death  2017     2017-12-08       104   
2  tt0060908             Le roi de coeur  2018     2018-06-08       102   
3  tt0069049  The Other Side of the Wind  2018     2018-11-02       122   
4  tt0071145                       Ankur  2019     2019-01-02       131   

             country worlwide_gross_income                 languages  \
0            Germany               $ 12156                    German   
1                 UK              $ 124241  English, French, Russian   
2      France, Italy               $ 18130   French, English, German   
3  France, Iran, USA                  None           English, German   
4              India                  None                     Hindi   

         production_company  
0          Decla-Bioscop AG  
1               The Archers  
2                 Fildebroc  
3  Royal Road Entertainment  
4    Blaze Film Enterprises

### Q1. Find the total number of rows in each table of the schema?

In [25]:
Query("select table_name,table_rows from information_schema.tables where table_schema='imdb';")

TABLE_NAME  TABLE_ROWS
0  director_mapping        3867
1             genre       14662
2             movie        7852
3             names       24806
4           ratings        7927

**we can use Alernative SQL Query to check**

In [65]:
Query("SELECT Count(*) FROM director_mapping")

|    |   Count(*) |
|---:|-----------:|
|  0 |       3867 |


In [64]:
Query("SELECT Count(*) FROM genre")

|    |   Count(*) |
|---:|-----------:|
|  0 |      14662 |


In [63]:
Query("SELECT Count(*) FROM movie")

|    |   Count(*) |
|---:|-----------:|
|  0 |       7997 |


In [62]:
Query("SELECT Count(*) FROM names")

|    |   Count(*) |
|---:|-----------:|
|  0 |      25735 |


In [61]:
Query("SELECT Count(*) FROM ratings")

|    |   Count(*) |
|---:|-----------:|
|  0 |       7997 |


### Q2. Which columns in the movie table have null values?

In [67]:
Query_df('''
SELECT
 sum(
  case
   WHEN id IS NULL THEN 1
   ELSE 0
  end
 ) AS ID_null,
 sum(
  case
   WHEN title IS NULL THEN 1
   ELSE 0
  end
 ) AS title_null,
 sum(
  case
   WHEN year IS NULL THEN 1
   ELSE 0
  end
 ) AS year_null,
 sum(
  case
   WHEN date_published IS NULL THEN 1
   ELSE 0
  end
 ) AS date_published_null,
 sum(
  case
   WHEN duration IS NULL THEN 1
   ELSE 0
  end
 ) AS duration_null,
 sum(
  case
   WHEN country IS NULL THEN 1
   ELSE 0
  end
 ) AS country_null,
 sum(
  case
   WHEN worlwide_gross_income IS NULL THEN 1
   ELSE 0
  end
 ) AS worlwide_gross_income_null,
 sum(
  case
   WHEN languages IS NULL THEN 1
   ELSE 0
  end
 ) AS language_null,
 sum(
  case
   WHEN production_company IS NULL THEN 1
   ELSE 0
  end
 ) AS production_company_null
FROM
 movie;
''')

ID_null title_null year_null date_published_null duration_null country_null  \
0       0          0         0                   0             0           20   

  worlwide_gross_income_null language_null production_company_null  
0                       3724           194                     528

### Q3. Find the total number of movies released each year? How does the trend look month wise? (Output expected)

In [ ]:
'''
/* Output format for the first part:
+---------------+-------------------+
| Year			|	number_of_movies|
+-------------------+----------------
|	2017		|	2134			|
|	2018		|		.			|
|	2019		|		.			|
+---------------+-------------------+
Output format for the second part of the question:
+---------------+-------------------+
|	month_num	|	number_of_movies|
+---------------+----------------
|	1			|	 134			|
|	2			|	 231			|
|	.			|		.			|
+---------------+-------------------+ */ 
'''

In [59]:
Query("SELECT year, count(title) as Number_of_movies FROM movie GROUP BY year")

|    |   year |   Number_of_movies |
|---:|-------:|-------------------:|
|  0 |   2017 |               3052 |
|  1 |   2018 |               2944 |
|  2 |   2019 |               2001 |


In [58]:
Query('''
SELECT
 MONTH(date_published) AS month,
 count(*) as number_of_movies
FROM
 movie
GROUP BY
 MONTH(date_published)
ORDER BY
 MONTH(date_published);
 ''')

|    |   month |   number_of_movies |
|---:|--------:|-------------------:|
|  0 |       1 |                804 |
|  1 |       2 |                640 |
|  2 |       3 |                824 |
|  3 |       4 |                680 |
|  4 |       5 |                625 |
|  5 |       6 |                580 |
|  6 |       7 |                493 |
|  7 |       8 |                678 |
|  8 |       9 |                809 |
|  9 |      10 |                801 |
| 10 |      11 |                625 |
| 11 |      12 |                438 |


**Highest Number of Movies are realeased in 2017**</br>
**The Highest number of movies is produced in the month of March**

###  Q4. How many movies were produced in the USA or India in the year 2019??

In [57]:
Query('''
SELECT
 Count(DISTINCT id) AS number_of_movies,
 year
FROM
 movie
WHERE
 (
  country LIKE '%INDIA%'
  OR country LIKE '%USA%'
 )
 AND year = 2019;
 ''')

|    |   number_of_movies |   year |
|---:|-------------------:|-------:|
|  0 |               1059 |   2019 |


**1059 Movies were produced in USA and India in the year 2019**

###  Q5. Find the unique list of the genres present in the data set?

In [56]:
Query('''
SELECT
 DISTINCT(genre)
FROM
 genre;
 ''')

|    | genre     |
|---:|:----------|
|  0 | Drama     |
|  1 | Fantasy   |
|  2 | Thriller  |
|  3 | Comedy    |
|  4 | Horror    |
|  5 | Family    |
|  6 | Romance   |
|  7 | Adventure |
|  8 | Action    |
|  9 | Sci-Fi    |
| 10 | Crime     |
| 11 | Mystery   |
| 12 | Others    |


**Movies belong to 13 genres in the dataset.**

### Q6.Which genre had the highest number of movies produced overall?

In [70]:
Query_df('''
SELECT
 genre,
 Count(m.id) AS number_of_movies
FROM
 movie AS m
 INNER JOIN genre AS g
WHERE
 g.movie_id = m.id
GROUP BY
 genre
ORDER BY
 number_of_movies DESC
LIMIT
 1;
''')

genre  number_of_movies
0  Drama              4285

**4265 Drama movies were produced in total and are the highest among all genres.**

### Q7. How many movies belong to only one genre?

In [75]:
Query_df('''
SELECT
 count(id) as movie_with_one_genre
FROM
 movie
WHERE
 id IN (
  SELECT
   movie_id
  FROM
   genre
  GROUP BY
   movie_id
  HAVING
   count(genre) = 1)
''')

movie_with_one_genre
0                  3289

**There is 3289 Movies with One Genre**

### Q8.What is the average duration of movies in each genre? 

In [ ]:
'''
+---------------+-------------------+
| genre			|	avg_duration	|
+-------------------+----------------
|	thriller	|		105			|
|	.			|		.			|
|	.			|		.			|
+---------------+-------------------+ 
'''

In [77]:
Query('''
SELECT
 genre,
 Round(Avg(duration), 2) AS avg_duration
FROM
 movie AS m
 INNER JOIN genre AS g ON g.movie_id = m.id
GROUP BY
 genre
ORDER BY
 avg_duration DESC;
 ''')

|    | genre     |   avg_duration |
|---:|:----------|---------------:|
|  0 | Action    |         112.88 |
|  1 | Romance   |         109.53 |
|  2 | Crime     |         107.05 |
|  3 | Drama     |         106.77 |
|  4 | Fantasy   |         105.14 |
|  5 | Comedy    |         102.62 |
|  6 | Adventure |         101.87 |
|  7 | Mystery   |         101.8  |
|  8 | Thriller  |         101.58 |
|  9 | Family    |         100.97 |
| 10 | Others    |         100.16 |
| 11 | Sci-Fi    |          97.94 |
| 12 | Horror    |          92.72 |


**Highest Duration goes to Action, Followed by Romance and Crime**

### Q9.What is the rank of the ‘thriller’ genre of movies among all the genres in terms of number of movies produced? 
- (Hint: Use the Rank function)

In [78]:
Query('''WITH a AS(
 SELECT
  genre,
  count(genre),
  rank() over(
   ORDER BY
    count(genre) DESC
  )
 FROM
  genre
 GROUP BY
  genre
)
SELECT
 *
FROM
 a
WHERE
 genre = 'thriller'
 ''')

|    | genre    |   count(genre) |            rank() over( |
|    |          |                |                ORDER BY |
|    |          |                |       count(genre) DESC |
|    |          |                |                       ) |
|---:|:---------|---------------:|------------------------:|
|  0 | Thriller |           1484 |                       3 |


**Thriller has a rank 3 and movie count of 1484**

### Q10.  Find the minimum and maximum values in  each column of the ratings table except the movie_id column?

In [ ]:
'''
/* Output format:
+---------------+-------------------+---------------------+----------------------+-----------------+-----------------+
| min_avg_rating|	max_avg_rating	|	min_total_votes   |	max_total_votes 	 |min_median_rating|min_median_rating|
+---------------+-------------------+---------------------+----------------------+-----------------+-----------------+
|		0		|			5		|	       177		  |	   2000	    		 |		0	       |	8			 |
+---------------+-------------------+---------------------+----------------------+-----------------+-----------------+*/
'''

In [81]:
Query_df('''
SELECT
 Min(avg_rating) AS MIN_AVG_RATING,
 Max(avg_rating) AS MAX_AVG_RATING,
 Min(total_votes) AS MIN_TOTAL_VOTES,
 Max(total_votes) AS MAX_TOTAL_VOTES,
 Min(median_rating) AS MIN_MEDIAN_RATING,
 Max(median_rating) AS MAX_MEDIAN_RATING
FROM
 ratings;
 ''')

MIN_AVG_RATING MAX_AVG_RATING  MIN_TOTAL_VOTES  MAX_TOTAL_VOTES  \
0            1.0           10.0              100           725138   

   MIN_MEDIAN_RATING  MAX_MEDIAN_RATING  
0                  1                 10

### Q11. Which are the top 10 movies based on average rating?

In [ ]:
'''
/* Output format:
+---------------+-------------------+---------------------+
| title			|		avg_rating	|		movie_rank    |
+---------------+-------------------+---------------------+
| Fan			|		9.6			|			5	  	  |
|	.			|		.			|			.		  |
|	.			|		.			|			.		  |
|	.			|		.			|			.		  |
+---------------+-------------------+---------------------+*/
'''

In [82]:
Query('''
SELECT
 m.title,
 r.avg_rating,
 dense_rank() over(
  ORDER BY
   r.avg_rating DESC
 ) AS m_rank
FROM
 movie m
 INNER JOIN ratings r ON (r.movie_id = m.id)
LIMIT
 10
 ''')

|    | title                          |   avg_rating |   m_rank |
|---:|:-------------------------------|-------------:|---------:|
|  0 | Kirket                         |         10   |        1 |
|  1 | Love in Kilnerry               |         10   |        1 |
|  2 | Gini Helida Kathe              |          9.8 |        2 |
|  3 | Runam                          |          9.7 |        3 |
|  4 | Fan                            |          9.6 |        4 |
|  5 | Android Kunjappan Version 5.25 |          9.6 |        4 |
|  6 | Yeh Suhaagraat Impossible      |          9.5 |        5 |
|  7 | Safe                           |          9.5 |        5 |
|  8 | The Brighton Miracle           |          9.5 |        5 |
|  9 | Shibu                          |          9.4 |        6 |


**Top 3 movies have a Average rating >= 9.8**

### Q12. Summarise the ratings table based on the movie counts by median ratings.

In [ ]:
'''
+---------------+-------------------+
| median_rating	|	movie_count		|
+-------------------+----------------
|	1			|		105			|
|	.			|		.			|
|	.			|		.			|
+---------------+-------------------+ */
-- Type your code below:
-- Order by is good to have
'''

In [83]:
Query('''
SELECT
 median_rating,
 Count(movie_id) AS movie_count
FROM
 ratings
GROUP BY
 median_rating
ORDER BY
 movie_count DESC;
 ''')

|    |   median_rating |   movie_count |
|---:|----------------:|--------------:|
|  0 |               7 |          2257 |
|  1 |               6 |          1975 |
|  2 |               8 |          1030 |
|  3 |               5 |           985 |
|  4 |               4 |           479 |
|  5 |               9 |           429 |
|  6 |              10 |           346 |
|  7 |               3 |           283 |
|  8 |               2 |           119 |
|  9 |               1 |            94 |


**Movies with a rating of 7 us Highest in number**

### Q13. Which production house has produced the most number of hit movies (average rating > 8)??

In [85]:
Query('''
WITH production_company_hit_movie_summary
     AS (SELECT production_company,
                Count(movie_id)                     AS MOVIE_COUNT,
                Rank()
                  OVER(
                    ORDER BY Count(movie_id) DESC ) AS PROD_COMPANY_RANK
         FROM   ratings AS R
                INNER JOIN movie AS M
                        ON M.id = R.movie_id
         WHERE  avg_rating > 8
                AND production_company IS NOT NULL
         GROUP  BY production_company)
SELECT *
FROM   production_company_hit_movie_summary
WHERE  prod_company_rank = 1; ''')

|    | production_company     |   MOVIE_COUNT |   PROD_COMPANY_RANK |
|---:|:-----------------------|--------------:|--------------------:|
|  0 | Dream Warrior Pictures |             3 |                   1 |
|  1 | National Theatre Live  |             3 |                   1 |


### Q14. How many movies released in each genre during March 2017 in the USA had more than 1,000 votes?

In [ ]:
'''
+---------------+-------------------+
| genre			|	movie_count		|
+-------------------+----------------
|	thriller	|		105			|
|	.			|		.			|
|	.			|		.			|
+---------------+-------------------+ */
-- Type your code below:

-- Query to find 
-- 1. Number of movies released in each genre 
-- 2. During March 2017 
-- 3. In the USA  (LIKE operator is used for pattern matching)
-- 4. Movies had more than 1,000 votes
'''

In [87]:
Query('''

SELECT
 genre,
 Count(M.id) AS MOVIE_COUNT
FROM
 movie AS M
 INNER JOIN genre AS G ON G.movie_id = M.id
 INNER JOIN ratings AS R ON R.movie_id = M.id
WHERE
 year = 2017
 AND MONTH(date_published) = 3
 AND country LIKE '%USA%'
 AND total_votes > 1000
GROUP BY
 genre
ORDER BY
 movie_count DESC;
 ''')

|    | genre     |   MOVIE_COUNT |
|---:|:----------|--------------:|
|  0 | Drama     |            24 |
|  1 | Comedy    |             9 |
|  2 | Action    |             8 |
|  3 | Thriller  |             8 |
|  4 | Sci-Fi    |             7 |
|  5 | Crime     |             6 |
|  6 | Horror    |             6 |
|  7 | Mystery   |             4 |
|  8 | Romance   |             4 |
|  9 | Fantasy   |             3 |
| 10 | Adventure |             3 |
| 11 | Family    |             1 |


- 24 Drama movies were released during March 2017 in the USA and had more than 1,000 votes
- Top 3 genres are drama, comedy and action during March 2017 in the USA and had more than 1,000 votes

###  Q15. Find movies of each genre that start with the word ‘The’ and which have an average rating > 8?

In [ ]:
'''
/* Output format:
+---------------+-------------------+---------------------+
| title			|		avg_rating	|		genre	      |
+---------------+-------------------+---------------------+
| Theeran		|		8.3			|		Thriller	  |
|	.			|		.			|			.		  |
|	.			|		.			|			.		  |
|	.			|		.			|			.		  |
+---------------+-------------------+---------------------+*/
-- Type your code below:

-- Query to find:
-- 1. Number of movies of each genre that start with the word ‘The’ (LIKE operator is used for pattern matching)
-- 2. Which have an average rating > 8?
-- Grouping by title to fetch distinct movie titles as movie belog to more than one genre
'''

In [88]:
Query('''
SELECT  title,
       avg_rating,
       genre
FROM   movie AS M
       INNER JOIN genre AS G
               ON G.movie_id = M.id
       INNER JOIN ratings AS R
               ON R.movie_id = M.id
WHERE  avg_rating > 8
       AND title LIKE 'THE%'
GROUP BY title
ORDER BY avg_rating DESC;
''')

|    | title                                |   avg_rating | genre   |
|---:|:-------------------------------------|-------------:|:--------|
|  0 | The Brighton Miracle                 |          9.5 | Drama   |
|  1 | The Colour of Darkness               |          9.1 | Drama   |
|  2 | The Blue Elephant 2                  |          8.8 | Drama   |
|  3 | The Irishman                         |          8.7 | Crime   |
|  4 | The Mystery of Godliness: The Sequel |          8.5 | Drama   |
|  5 | The Gambinos                         |          8.4 | Crime   |
|  6 | Theeran Adhigaaram Ondru             |          8.3 | Action  |
|  7 | The King and I                       |          8.2 | Drama   |


- There are 8 movies which begin with "The" in their title.
- The Brighton Miracle has highest average rating of 9.5.
- All the movies belong to the top 3 genres.

### Q16. Of the movies released between 1 April 2018 and 1 April 2019, how many were given a median rating of 8?

In [89]:
Query('''
SELECT
 median_rating,
 Count(*) AS movie_count
FROM
 movie AS M
 INNER JOIN ratings AS R ON R.movie_id = M.id
WHERE
 median_rating = 8
 AND date_published BETWEEN '2018-04-01'
 AND '2019-04-01'
GROUP BY
 median_rating;
 ''')

|    |   median_rating |   movie_count |
|---:|----------------:|--------------:|
|  0 |               8 |           361 |


###  Q17. Do German movies get more votes than Italian movies? 

In [91]:
Query('''
SELECT
 country,
 sum(total_votes) AS total_votes
FROM
 movie AS m
 INNER JOIN ratings AS r ON m.id = r.movie_id
WHERE
 country = 'Germany'
 OR country = 'Italy'
GROUP BY
 country;
''')

|    | country   |   total_votes |
|---:|:----------|--------------:|
|  0 | Germany   |        106710 |
|  1 | Italy     |         77965 |


**Yes!! German Movies get more votes than Italian Movies**

### Q18. Which columns in the names table have null values??

In [ ]:
'''/*Hint: You can find null values for individual columns or follow below output format
+---------------+-------------------+---------------------+----------------------+
| name_nulls	|	height_nulls	|date_of_birth_nulls  |known_for_movies_nulls|
+---------------+-------------------+---------------------+----------------------+
|		0		|			123		|	       1234		  |	   12345	    	 |
+---------------+-------------------+---------------------+----------------------+*/'''

In [92]:
Query('''
SELECT
 count(*) - count(name) AS name_nulls,
 count(*) - count(height) AS height_nulls,
 count(*) - count(date_of_birth) AS date_of_birth_nulls,
 count(*) - count(known_for_movies) AS known_for_movies_nulls
FROM
 NAMES
 ''')

|    |   name_nulls |   height_nulls |   date_of_birth_nulls |   known_for_movies_nulls |
|---:|-------------:|---------------:|----------------------:|-------------------------:|
|  0 |            0 |          17335 |                 13431 |                    15226 |


### Q19. Who are the top three directors in the top three genres whose movies have an average rating > 8?

In [ ]:
'''
/* Output format:
+---------------+-------------------+
| director_name	|	movie_count		|
+---------------+-------------------|
|James Mangold	|		4			|
|	.			|		.			|
|	.			|		.			|
+---------------+-------------------+ */
'''

In [93]:
Query('''
WITH top3genre AS (
 SELECT
  g.genre
 FROM
  movie m
  JOIN ratings r ON m.id = r.movie_id
  JOIN genre g ON m.id = g.movie_id
 WHERE
  avg_rating > 8
 GROUP BY
  genre
 ORDER BY
  count(title) DESC
 LIMIT
  3
)
SELECT
 name AS "Director_name",
 count(g2.movie_id) AS "movie_count"
FROM
 director_mapping dm
 JOIN NAMES n ON n.id = dm.name_id
 JOIN genre g2 ON g2.movie_id = dm.movie_id
 JOIN ratings r ON r.movie_id = dm.movie_id
 JOIN top3genre tg ON g2.genre = tg.genre
WHERE
 r.avg_rating > 8
GROUP BY
 name
ORDER BY
 count(dm.movie_id) DESC
LIMIT
 3;
 ''')

|    | Director_name   |   movie_count |
|---:|:----------------|--------------:|
|  0 | James Mangold   |             4 |
|  1 | Anthony Russo   |             3 |
|  2 | Soubin Shahir   |             3 |


**James Mangold , Joe Russo and Anthony Russo are top three directors in the top three genres whose movies have an average rating > 8**

### Q20. Who are the top two actors whose movies have a median rating >= 8?

In [ ]:
'''
/* Output format:
+---------------+-------------------+
| actor_name	|	movie_count		|
+-------------------+----------------
|Christain Bale	|		10			|
|	.			|		.			|
+---------------+-------------------+ */
'''

In [94]:
Query('''
SELECT
 DISTINCT name,
 COUNT(r.movie_id)
FROM
 ratings AS r
 INNER JOIN role_mapping AS rm ON (rm.movie_id = r.movie_id)
 INNER JOIN NAMES AS n ON (rm.name_id = n.id)
WHERE
 (r.median_rating >= 8)
 AND rm.category = 'actor'
GROUP BY
 name
ORDER BY
 r.movie_id
LIMIT
 2;
''')

|    | name         |   COUNT(r.movie_id) |
|---:|:-------------|--------------------:|
|  0 | David Niven  |                   1 |
|  1 | Robert Coote |                   1 |


###  Q21. Which are the top three production houses based on the number of votes received by their movies?

In [ ]:
'''
/* Output format:
+------------------+--------------------+---------------------+
|production_company|vote_count			|		prod_comp_rank|
+------------------+--------------------+---------------------+
| The Archers		|		830			|		1	  		  |
|	.				|		.			|			.		  |
|	.				|		.			|			.		  |
+-------------------+-------------------+---------------------+*/
-- Type your code below:
'''

In [95]:
Query_df('''
SELECT
 production_company,
 Sum(total_votes) AS vote_count,
 Rank() OVER(
  ORDER BY
   Sum(total_votes) DESC
 ) AS prod_comp_rank
FROM
 movie AS m
 INNER JOIN ratings AS r ON r.movie_id = m.id
GROUP BY
 production_company
LIMIT
 3;
 ''')

production_company vote_count  prod_comp_rank
0         Marvel Studios    2656967               1
1  Twentieth Century Fox    2411163               2
2           Warner Bros.    2396057               3

**Top three production houses based on the number of votes received by their movies are Marvel Studios, Twentieth Century Fox and Warner Bros.**

### Q22. Rank actors with movies released in India based on their average ratings. Which actor is at the top of the list?

In [ ]:
'''
/* Output format:
+---------------+-------------------+---------------------+----------------------+-----------------+
| actor_name	|	total_votes		|	movie_count		  |	actor_avg_rating 	 |actor_rank	   |
+---------------+-------------------+---------------------+----------------------+-----------------+
|	Yogi Babu	|			3455	|	       11		  |	   8.42	    		 |		1	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
+---------------+-------------------+---------------------+----------------------+-----------------+*/
-- Type your code below:'''

In [96]:
Query('''

WITH actor_summary AS (
 SELECT
  N.NAME AS actor_name,
  total_votes,
  Count(R.movie_id) AS movie_count,
  Round(
   Sum(avg_rating * total_votes) / Sum(total_votes),
   2
  ) AS actor_avg_rating
 FROM
  movie AS M
  INNER JOIN ratings AS R ON M.id = R.movie_id
  INNER JOIN role_mapping AS RM ON M.id = RM.movie_id
  INNER JOIN NAMES AS N ON RM.name_id = N.id
 WHERE
  category = 'ACTOR'
  AND country = "india"
 GROUP BY
  NAME
 HAVING
  movie_count >= 5
)
SELECT
 *,
 Rank() OVER(
  ORDER BY
   actor_avg_rating DESC
 ) AS actor_rank
FROM
 actor_summary;
''')

|    | actor_name         |   total_votes |   movie_count |   actor_avg_rating |   actor_rank |
|---:|:-------------------|--------------:|--------------:|-------------------:|-------------:|
|  0 | Vijay Sethupathi   |         20364 |             5 |               8.42 |            1 |
|  1 | Fahadh Faasil      |          3684 |             5 |               7.99 |            2 |
|  2 | Yogi Babu          |           223 |            11 |               7.83 |            3 |
|  3 | Joju George        |           413 |             5 |               7.58 |            4 |
|  4 | Ammy Virk          |           169 |             6 |               7.55 |            5 |
|  5 | Dileesh Pothan     |          1115 |             5 |               7.52 |            6 |
|  6 | Kunchacko Boban    |          3684 |             6 |               7.48 |            7 |
|  7 | Pankaj Tripathi    |         13723 |             5 |               7.44 |            8 |
|  8 | Rajkummar Rao      |          832

**Top actor is Vijay Sethupathi followed by Fahadh Faasil and Yogi Babu.**

###  Q23.Find out the top five actresses in Hindi movies released in India based on their average ratings? 

In [ ]:
'''
/* Output format:
+---------------+-------------------+---------------------+----------------------+-----------------+
| actress_name	|	total_votes		|	movie_count		  |	actress_avg_rating 	 |actress_rank	   |
+---------------+-------------------+---------------------+----------------------+-----------------+
|	Tabu		|			3455	|	       11		  |	   8.42	    		 |		1	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
+---------------+-------------------+---------------------+----------------------+-----------------+*/'''

In [97]:
Query('''

SELECT
 name,
 total_votes,
 count(m.id) AS m_count,
 SUM(r.avg_rating * r.total_votes) / SUM(r.total_votes) AS a_avg_rating,
 dense_rank() over(
  ORDER BY
   avg_rating DESC,
   total_votes DESC
 ) AS actress_rank
FROM
 movie m
 JOIN ratings r ON m.id = r.movie_id
 JOIN role_mapping rm ON rm.movie_id = m.id
 JOIN NAMES n ON rm.name_id = n.id
WHERE
 m.country = 'India'
 AND rm.category = 'actress'
GROUP BY
 name
HAVING
 count(m.title) >= 3;''')

|    | name                  |   total_votes |   m_count |   a_avg_rating |   actress_rank |
|---:|:----------------------|--------------:|----------:|---------------:|---------------:|
|  0 | Anusree               |          1017 |         4 |        7.56253 |              1 |
|  1 | Anupama Parameshwaran |           613 |         3 |        7.4119  |              2 |
|  2 | Regina Cassandra      |          1793 |         4 |        7.7464  |              3 |
|  3 | Nirmal Rishi          |           552 |         4 |        7.29145 |              4 |
|  4 | Kajal Aggarwal        |         22704 |         3 |        7.13767 |              5 |
|  5 | Aparna Balamurali     |          1197 |         3 |        7.31803 |              6 |
|  6 | Kriti Sanon           |         14978 |         3 |        7.04911 |              7 |
|  7 | Manju Warrier         |          5471 |         5 |        6.76135 |              8 |
|  8 | Rachita Ram           |           242 |         3 |        6.95

###  Q24. Select thriller movies as per avg rating and classify them in the following category: 

In [98]:
Query('''
SELECT
 m.title,
 case
  WHEN avg_rating > 8 THEN 'Superhit movies'
  WHEN avg_rating BETWEEN 7
  AND 8 THEN 'Hit movies'
  WHEN avg_rating BETWEEN 5
  AND 7 THEN 'One-time-watch movies'
  WHEN avg_rating < 5 THEN 'Flop movies'
 end AS Rating
FROM
 movie m
 INNER JOIN genre g ON (m.id = g.movie_id)
 INNER JOIN ratings r ON (g.movie_id = r.movie_id)
WHERE
 g.genre = 'thriller'
 ''')

|      | title                                                         | Rating                |
|-----:|:--------------------------------------------------------------|:----------------------|
|    0 | Der müde Tod                                                  | Hit movies            |
|    1 | Fahrenheit 451                                                | Flop movies           |
|    2 | Pet Sematary                                                  | One-time-watch movies |
|    3 | Dukun                                                         | One-time-watch movies |
|    4 | Back Roads                                                    | Hit movies            |
|    5 | Countdown                                                     | One-time-watch movies |
|    6 | Staged Killer                                                 | Flop movies           |
|    7 | Vellaipookal                                                  | Hit movies            |
|    8 | Uriyadi 2            

### Q25. What is the genre-wise running total and moving average of the average movie duration? 

In [101]:
Query('''
SELECT
 genre,
 ROUND(AVG(duration), 2) AS avg_duration,
 SUM(ROUND(AVG(duration), 2)) OVER(
  ORDER BY
   genre ROWS UNBOUNDED PRECEDING
 ) AS running_total_duration,
 AVG(ROUND(AVG(duration), 2)) OVER(
  ORDER BY
   genre ROWS 10 PRECEDING
 ) AS moving_avg_duration
FROM
 movie AS m
 INNER JOIN genre AS g ON m.id = g.movie_id
GROUP BY
 genre
ORDER BY
 genre;
 ''')

|    | genre     |   avg_duration |   running_total_duration |   moving_avg_duration |
|---:|:----------|---------------:|-------------------------:|----------------------:|
|  0 | Action    |         112.88 |                   112.88 |               112.88  |
|  1 | Adventure |         101.87 |                   214.75 |               107.375 |
|  2 | Comedy    |         102.62 |                   317.37 |               105.79  |
|  3 | Crime     |         107.05 |                   424.42 |               106.105 |
|  4 | Drama     |         106.77 |                   531.19 |               106.238 |
|  5 | Family    |         100.97 |                   632.16 |               105.36  |
|  6 | Fantasy   |         105.14 |                   737.3  |               105.329 |
|  7 | Horror    |          92.72 |                   830.02 |               103.752 |
|  8 | Mystery   |         101.8  |                   931.82 |               103.536 |
|  9 | Others    |         100.16 |        

###  Q26. Which are the five highest-grossing movies of each year that belong to the top three genres? 

In [ ]:
'''
/* Output format:
+---------------+-------------------+---------------------+----------------------+-----------------+
| genre			|	year			|	movie_name		  |worldwide_gross_income|movie_rank	   |
+---------------+-------------------+---------------------+----------------------+-----------------+
|	comedy		|			2017	|	       indian	  |	   $103244842	     |		1	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
+---------------+-------------------+---------------------+----------------------+-----------------+*/
-- Type your code below:

-- Top 3 Genres based on most number of movies'''

In [102]:
Query('''

WITH top_3 AS(
 SELECT
  g.genre,
  count(r.movie_id)
 FROM
  genre g
  INNER JOIN ratings r ON (g.movie_id = r.movie_id)
 GROUP BY
  g.genre
 ORDER BY
  count(r.movie_id) DESC
 LIMIT
  3
), top5 AS (
 SELECT
  g.genre,
  m.title,
  m.year,
  m.worlwide_gross_income,
  dense_rank() over(
   PARTITION by m.year
   ORDER BY
    m.worlwide_gross_income DESC
  ) AS movie_rank
 FROM
  movie m
  INNER JOIN genre g ON (g.movie_id = m.id)
 WHERE
  g.genre IN (
   SELECT
    genre
   FROM
    top_3
  )
)
SELECT
 *
FROM
 top5
WHERE
 movie_rank <= 5
''')

|    | genre    | title                      |   year | worlwide_gross_income   |   movie_rank |
|---:|:---------|:---------------------------|-------:|:------------------------|-------------:|
|  0 | Drama    | Shatamanam Bhavati         |   2017 | INR 530500000           |            1 |
|  1 | Drama    | Winner                     |   2017 | INR 250000000           |            2 |
|  2 | Drama    | Thank You for Your Service |   2017 | $ 9995692               |            3 |
|  3 | Comedy   | The Healer                 |   2017 | $ 9979800               |            4 |
|  4 | Drama    | The Healer                 |   2017 | $ 9979800               |            4 |
|  5 | Thriller | Gi-eok-ui bam              |   2017 | $ 9968972               |            5 |
|  6 | Thriller | The Villain                |   2018 | INR 1300000000          |            1 |
|  7 | Drama    | Antony & Cleopatra         |   2018 | $ 998079                |            2 |
|  8 | Comedy   | La fuitina s

###  Q27.  Which are the top two production houses that have produced the highest number of hits (median rating >= 8) among multilingual movies?


In [ ]:
'''
/* Output format:
+-------------------+-------------------+---------------------+
|production_company |movie_count		|		prod_comp_rank|
+-------------------+-------------------+---------------------+
| The Archers		|		830			|		1	  		  |
|	.				|		.			|			.		  |
|	.				|		.			|			.		  |
+-------------------+-------------------+---------------------+*/
-- Type your code below:
'''

In [104]:
Query('''

SELECT
 m.production_company,
 dense_rank() over(
  ORDER BY
   r.movie_id DESC
 ) AS pro_rank
FROM
 movie m
 INNER JOIN ratings r ON(r.movie_id = m.id)
WHERE
 r.median_rating >= 8
 AND position(',' IN languages) > 0
 AND production_company IS NOT NULL
GROUP BY
 production_company
LIMIT
 2
 ''')

|    | production_company   |   pro_rank |
|---:|:---------------------|-----------:|
|  0 | Jorkwang Films       |          1 |
|  1 | MF Production        |          2 |


### Q28. Who are the top 3 actresses based on number of Super Hit movies (average rating >8) in drama genre?

In [ ]:
'''
/* Output format:
+---------------+-------------------+---------------------+----------------------+-----------------+
| actress_name	|	total_votes		|	movie_count		  |actress_avg_rating	 |actress_rank	   |
+---------------+-------------------+---------------------+----------------------+-----------------+
|	Laura Dern	|			1016	|	       1		  |	   9.60			     |		1	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
|		.		|			.		|	       .		  |	   .	    		 |		.	       |
+---------------+-------------------+---------------------+----------------------+-----------------+*/'''

In [106]:
Query_df('''

WITH actress_summary AS (
 SELECT
  n.NAME AS actress_name,
  SUM(total_votes) AS total_votes,
  Count(r.movie_id) AS movie_count,
  Round(Sum(avg_rating * total_votes) / Sum(total_votes), 2) AS actress_avg_rating
 FROM
  movie AS m
  INNER JOIN ratings AS r ON m.id = r.movie_id
  INNER JOIN role_mapping AS rm ON m.id = rm.movie_id
  INNER JOIN NAMES AS n ON rm.name_id = n.id
  INNER JOIN GENRE AS g ON g.movie_id = m.id
 WHERE
  category = 'ACTRESS'
  AND avg_rating > 8
  AND genre = "Drama"
 GROUP BY
  NAME
)
SELECT
 *,
 Rank() OVER(
  ORDER BY
   movie_count DESC
 ) AS actress_rank
FROM
 actress_summary
LIMIT
 3;
 ''')

actress_name total_votes  movie_count actress_avg_rating  \
0  Parvathy Thiruvothu        4974            2               8.25   
1          Susan Brown         656            2               8.94   
2      Amanda Lawrence         656            2               8.94   

   actress_rank  
0             1  
1             1  
2             1

**Top 3 actresses based on number of Super Hit movies are Parvathy Thiruvothu, Susan Brown and Amanda Lawrence**